gh-119786: added InternalDocs/generators.md (#128524)
python/cpython: InternalDocs/generators.md

Generators and Coroutines
=========================

Generators
----------

> The implementation of generators in CPython consists of instances of `PyGenObject`

Reviewer: ```suggestion
Generators in CPython are implemented with the struct `PyGenObject`. Generators consist of a frame and some metadata.
```

`PyGenObject` is a C struct, not a Python type.

> and bytecode instructions that operate on instances of this type.

Reviewer: I'd just delete this. Generator objects only reference their bytecode very indirectly (through the frame and then the code object)

> A generator object is invoked in a [`frame`](frames.md), like a function.

Reviewer: What do you mean "invoked in a frame".

If you want to make parallels with functions:
* functions are called, we *send* to a generator 
* functions return, generators yield (and sometimes return, but mainly yield)

in the sense that it puts a value on the stack and returns execution to the
calling frame, but it also needs to perform additional work to leave the generator
frame in a state that allows it to be resumed.  In particular, it updates the frame's
instruction pointer and stores the interpreter's exception state on the generator
object. When the generator is resumed, this exception state is copied back to the
interpreter state.

The `frame` of a generator is embedded in the generator object struct as a
[`_PyInterpreterFrame`](frames.md) (see `_PyGenObject_HEAD` in
[`pycore_genobject.h`](../Include/internal/pycore_genobject.h)).
This means that we can get the frame from the generator or the generator
from the frame (see `_PyGen_GetGeneratorFromFrame` in the same file).
Other fields of the generator struct include metadata (such as the name of
the generator function) and runtime state information (such as whether its
frame is executing, suspended, cleared, etc.).

Generator Object Creation and Destruction
-----------------------------------------

The bytecode of a generator function begins with a
[`RETURN_GENERATOR`](https://docs.python.org/dev/library/dis.html#opcode-RETURN_GENERATOR)

> instruction, which creates a generator object, along with its embedded frame.

Reviewer: ```suggestion
instruction, which creates a generator object, including its embedded frame.
```
"along with" suggests the frame is external to the generator

The generator's frame is initialized as a copy of the frame in which
`RETURN_GENERATOR` is executing, but its `owner` field is overwritten to indicate
that it is owned by a generator. Finally, `RETURN_GENERATOR` pushes the new generator
object to the stack and returns to the caller of the generator function. When the
generator is next resumed by [`gen_send_ex2()`](../Objects/genobject.c),
`_PyEval_EvalFrame()` is called to continue executing the generator function,

> in the frame that is embedded in the generator object.

Reviewer: Maybe mention that the prototype frame is destroyed in this process.

> When a generator object is destructed in [`gen_dealloc`](../Objects/genobject.c),

Reviewer: ```suggestion
When a generator object is destroyed in [`gen_dealloc`](../Objects/genobject.c),
```

its embedded `_PyInterpreterFrame` field may need to be preserved, if it is exposed
to Python as part of a [`PyFrameObject`](frames.md#frame-objects). This is detected
in [`_PyFrame_ClearExceptCode`](../Python/frame.c) by the fact that the interpreter
frame's `frame_obj` field is set, and the frame object it points to has refcount
greater than 1. If so, the `take_ownership()` function is called to create a new
copy of the interpreter frame and transfer ownership of it from the generator to
the frame object.

Iteration
---------

The [`FOR_ITER`](https://docs.python.org/dev/library/dis.html#opcode-FOR_ITER)
instruction calls `__next__` on the iterator which is on the top of the stack,
and pushes the result to the stack. It has [`specializations`](adaptive.md)
for a few common iterator types, including `FOR_ITER_GEN`, for iterating over
a generator.

> 

Reviewer: I think it is worth mentioning that `FOR_ITER_GEN` works by pushing the generator's frame and jumping to the next instruction after the last yield, avoiding the call to the C implementation of `__next__`.

Generators and Coroutines
=========================

Generators
----------

> The implementation of generators in CPython consists of the builtin object type

Reviewer: Rather than say "consist of `PyGenObject`", I'd say "instances of `PyGenObject`".

Logically (and in the implementation until we inlined frames) a generator object consists of just a pointer to a frame object and some metadata (its name and whether it is currently executing, suspended or closed)

`PyGenObject` and bytecode instructions that operate on instances of this type.

> A generator object executes in its own [`frame`](frames.md), like a function.

Reviewer: I don't think of frames being owned by a function.
There is no reference to a frame object from a function object.

The difference is that a function returns only once, while a generator
"returns" to the caller every time it emits a new item with a
[`yield` expression](https://docs.python.org/dev/reference/expressions.html#yield-expressions).
This is implemented by the
[`YIELD_VALUE`](https://docs.python.org/dev/library/dis.html#opcode-YIELD_VALUE)
bytecode, which is similar to
[`RETURN_VALUE`](https://docs.python.org/dev/library/dis.html#opcode-RETURN_VALUE)
in the sense that it puts a value on the stack and returns execution to the
calling frame, but it also needs to perform additional work to leave the generator
frame in a state that allows it to be resumed.  In particular, it updates the frame's
instruction pointer and stores the interpreter's exception state on the generator

> object. When the generator it resumed, this exception state is copied back to the

Reviewer 2: ```suggestion
object. When the generator is resumed, this exception state is copied back to the
```
